motr/ut/protection_info_checks.c

```diff
@@ -0,0 +1,327 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#ifndef __KERNEL__
+#include <openssl/md5.h>
+#endif /* __KERNEL__ */
+
+#include "ut/ut.h"
+#include "motr/client.h"
+#include "motr/client_internal.h"
+#include "motr/ut/client.h"
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_CLIENT
+#include "lib/trace.h"          /* M0_LOG */
+
+struct m0_ut_suite ut_suite_pi;
+
+enum {
+	BUFFER_SIZE = 4096,
+	SEGS_NR     = 16,
+
+	BIG_BUFFER_SIZE = 4096*16,
+	BIG_SEGS_NR = 10
+};
+
+#define DATA_UNIT_COUNT 10
+#define OBJ_CONTAINER 0x123
+#define OBJ_KEY 0x456
+
+struct m0_bufvec *user_data;
+unsigned char *curr_context[DATA_UNIT_COUNT];
+unsigned char *seeded_sum[DATA_UNIT_COUNT];
+unsigned char *final_sum;
+
+struct m0_bufvec *big_user_data;
+unsigned char *big_curr_context;
+unsigned char *big_final_sum;
+
+M0_INTERNAL int pi_init(void)
+{
+#ifndef __KERNEL__
+	ut_shuffle_test_order(&ut_suite_pi);
+#endif
+
+	int i, j, rc;
+	// allocate and populate buffers
+
+	M0_ALLOC_ARR(user_data, DATA_UNIT_COUNT);
+	M0_UT_ASSERT(user_data != NULL);
+
+	for (j = 0; j < DATA_UNIT_COUNT; j++) {
+		rc = m0_bufvec_alloc(&user_data[j], SEGS_NR, BUFFER_SIZE);
+		M0_UT_ASSERT(rc == 0);
+		for (i = 0; i < user_data[j].ov_vec.v_nr; ++i) {
+			memset(user_data[j].ov_buf[i], 'a' + j, BUFFER_SIZE);
+		}
+
+		M0_ALLOC_ARR(curr_context[j], sizeof(MD5_CTX));
+		M0_UT_ASSERT(curr_context[j] != NULL);
+		M0_ALLOC_ARR(seeded_sum[j], MD5_DIGEST_LENGTH);
+		M0_UT_ASSERT(seeded_sum[j] != NULL);
+	}
+
+	M0_ALLOC_ARR(final_sum, MD5_DIGEST_LENGTH);
+	M0_UT_ASSERT(final_sum != NULL);
+
+	M0_ALLOC_ARR(big_user_data, 1);
+	M0_UT_ASSERT(big_user_data != NULL);
+	rc = m0_bufvec_alloc(big_user_data, BIG_SEGS_NR, BIG_BUFFER_SIZE);
+	M0_UT_ASSERT(rc == 0);
+	for (i = 0; i < big_user_data->ov_vec.v_nr; ++i) {
+		memset(big_user_data->ov_buf[i], 'a' + i, BIG_BUFFER_SIZE);
+	}
+
+	M0_ALLOC_ARR(big_curr_context, sizeof(MD5_CTX));
+	M0_UT_ASSERT(big_curr_context != NULL);
+
+	M0_ALLOC_ARR(big_final_sum, MD5_DIGEST_LENGTH);
+	M0_UT_ASSERT(big_final_sum != NULL);
+
+	return 0;
+}
+
+M0_INTERNAL int pi_fini(void)
+{
+	int j;
+
+	for (j = 0; j < DATA_UNIT_COUNT; j++) {
+		m0_bufvec_free(&user_data[j]);
+		m0_free(seeded_sum[j]);
+		m0_free(curr_context[j]);
+	}
+
+	m0_free(user_data);
+	m0_free(final_sum);
+	m0_free(big_curr_context);
+	m0_free(big_final_sum);
+	m0_bufvec_free(big_user_data);
+	m0_free(big_user_data);
+	return 0;
+}
+
+void verify_case_one_two(void)
+{
+	int i, j, rc;
+	MD5_CTX curr_ctx;
+	MD5_CTX tmp_ctx;
+	char seed_str[64] = {'\0'};
+	struct m0_pi_seed seed;
+	unsigned char *seed_sum = (unsigned char *)malloc(MD5_DIGEST_LENGTH);
+	unsigned char *unseed_final_sum =
+		(unsigned char *)malloc(MD5_DIGEST_LENGTH);
```

Comment:
Use M0_ALLOC_ARR() to allocate memory for seed_sum and unseed_final_sum and verify the same with M0_UT_ASSERT.

Reply:
Done. Thanks for catching this.